File: Engine/src/main/java/org/etieskrill/engine/graphics/animation/BoneAnimation.java
package org.etieskrill.engine.graphics.animation;

import org.etieskrill.engine.graphics.model.Bone;
import org.joml.Quaternionfc;
import org.joml.Vector3fc;

import java.util.List;

public record BoneAnimation(
        Bone bone,
        List<Vector3fc> positions,
        List<Double> positionTimes,
        List<Quaternionfc> rotations,
        List<Double> rotationTimes,
        List<Vector3fc> scalings,
        List<Double> scaleTimes,
        Animation.Behaviour preBehaviour, //what happens before first key
        Animation.Behaviour postBehaviour //what happens after last key
) {
    @Override
    public String toString() {
        return "BoneAnimation{" +
                "bone=" + bone +
                ", num positions=" + positions.size() +
                ", num positionTimes=" + positionTimes.size() +
                ", num rotations=" + rotations.size() +
                ", num rotationTimes=" + rotationTimes.size() +
                ", num scalings=" + scalings.size() +
                ", num scaleTimes=" + scaleTimes.size() +
                ", preBehaviour=" + preBehaviour +
                ", postBehaviour=" + postBehaviour +
                '}';
    }
}

File: Engine/src/commonMain/kotlin/org/etieskrill/engine/graphics/animation/BoneAnimation.kt
package org.etieskrill.engine.graphics.animation

import org.etieskrill.engine.graphics.model.Bone
import org.joml.Quaternionfc
import org.joml.Vector3fc

data class BoneAnimation(
    val bone: Bone,
    val positions: List<Vector3fc>,
    val positionTimes: DoubleArray,
    val rotations: List<Quaternionfc>,
    val rotationTimes: DoubleArray,
    val scalings: List<Vector3fc>,
    val scaleTimes: DoubleArray,
    val preBehaviour: AnimationBehaviour, //what happens before first key
    val postBehaviour: AnimationBehaviour //what happens after last key
) {
    override fun toString(): String {
        return "BoneAnimation(bone=$bone, positions=$positions, positionTimes=${
            positionTimes.size
        }, rotations=$rotations, rotationTimes=${rotationTimes.size}, scalings=$scalings, scaleTimes=${
            scaleTimes.size
        }, preBehaviour=$preBehaviour, postBehaviour=$postBehaviour)"
    }
}
